### PR TITLE
Add QueryParameters option to AWS CloudMap-based peer discovery

### PR DESCRIPTION
**Pull request by engechas**
### Description
This PR adds `awsCloudMapQueryParameters` as an optional setting when using AWS CloudMap for peer discovery. The implementation accepts a `Map<String, String>` which is passed directly to the `queryParameters` field of the `DiscoverInstancesRequest`.

Documentation on this field: https://sdk.amazonaws.com/java/api/latest/software/amazon/awssdk/services/servicediscovery/model/DiscoverInstancesRequest.Builder.html#queryParameters-java.util.Map-
 
### Testing
AWS CloudMap service contents:
```
aws servicediscovery --region eu-west-1 discover-instances --namespace serv-disc --service serv-disc
{
    "Instances": [
        {
            "InstanceId": "a2ae948ed1244897bf8dfa404fc9593c",
            "NamespaceName": "serv-disc",
            "ServiceName": "serv-disc",
            "HealthStatus": "HEALTHY",
            "Attributes": {
                "AWS_INSTANCE_IPV4": "<redacted>",
                "REGION": "eu-west-1"
            }
        },
        {
            "InstanceId": "30e368d910bf421e98a246d7275987ba",
            "NamespaceName": "serv-disc",
            "ServiceName": "serv-disc",
            "HealthStatus": "HEALTHY",
            "Attributes": {
                "AWS_INSTANCE_IPV4": "<redacted>",
                "REGION": "eu-west-1",
                "test": "filter-params"
            }
        },
        {
            "InstanceId": "c18c00a21852483ebefbb020510bb910",
            "NamespaceName": "serv-disc",
            "ServiceName": "serv-disc",
            "HealthStatus": "HEALTHY",
            "Attributes": {
                "AWS_INSTANCE_IPV4": "<redacted>",
                "REGION": "eu-west-1"
            }
        }
    ]
}
```
#### Testing no filters
pipelines.yaml:
```
simple-sample-pipeline:
  workers: 2
  delay: "5000"
  source:
    random:
  processor:
    - peer_forwarder:
        discovery_mode: aws_cloud_map
        ssl: false
        awsCloudMapNamespaceName: serv-disc
        awsCloudMapServiceName: serv-disc
        awsRegion: eu-west-1
  sink:
    - stdout:
```
Logs:
```
2022-07-01T22:41:13,504 [sdk-async-response-0-0] INFO  com.amazon.dataprepper.plugins.prepper.peerforwarder.discovery.AwsCloudMapPeerListProvider$AwsCloudMapDynamicEndpointGroup - Discovered 3 instances.
```
#### Testing filter has match:
pipelines.yaml:
```
simple-sample-pipeline:
  workers: 2
  delay: "5000"
  source:
    random:
  processor:
    - peer_forwarder:
        discovery_mode: aws_cloud_map
        ssl: false
        awsCloudMapNamespaceName: serv-disc
        awsCloudMapServiceName: serv-disc
        awsRegion: eu-west-1
        awsCloudMapQueryParameters:
          test: filter-params
  sink:
    - stdout:
```
Logs:
```
2022-07-01T22:58:23,312 [sdk-async-response-0-0] INFO  com.amazon.dataprepper.plugins.prepper.peerforwarder.discovery.AwsCloudMapPeerListProvider$AwsCloudMapDynamicEndpointGroup - Discovered 1 instances.
```
#### Testing filter has no matches
pipelines.yaml:
```
simple-sample-pipeline:
  workers: 2
  delay: "5000"
  source:
    random:
  processor:
    - peer_forwarder:
        discovery_mode: aws_cloud_map
        ssl: false
        awsCloudMapNamespaceName: serv-disc
        awsCloudMapServiceName: serv-disc
        awsRegion: eu-west-1
        awsCloudMapQueryParameters:
          test: no-match
  sink:
    - stdout:
```
Logs:
```
2022-07-01T22:59:47,138 [sdk-async-response-0-0] INFO  com.amazon.dataprepper.plugins.prepper.peerforwarder.discovery.AwsCloudMapPeerListProvider$AwsCloudMapDynamicEndpointGroup - Discovered 0 instances.
2022-07-01T22:59:47,141 [sdk-async-response-0-0] INFO  com.amazon.dataprepper.plugins.prepper.peerforwarder.HashRing - Building hash ring with endpoints: []
```
#### Testing blank entry
pipelines.yaml:
```
simple-sample-pipeline:
  workers: 2
  delay: "5000"
  source:
    random:
  processor:
    - peer_forwarder:
        discovery_mode: aws_cloud_map
        ssl: false
        awsCloudMapNamespaceName: serv-disc
        awsCloudMapServiceName: serv-disc
        awsRegion: eu-west-1
        awsCloudMapQueryParameters:
  sink:
    - stdout:
```
Logs:
```
2022-07-01T23:02:33,129 [sdk-async-response-0-0] INFO  com.amazon.dataprepper.plugins.prepper.peerforwarder.discovery.AwsCloudMapPeerListProvider$AwsCloudMapDynamicEndpointGroup - Discovered 3 instances.
```

### Check List
- [x] New functionality includes testing.
- [x] New functionality has been documented.
  - [N/A] New functionality has javadoc added
- [x] Commits are signed with a real name per the DCO

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/data-prepper/blob/main/CONTRIBUTING.md).

### CODE REVIEW

File: data-prepper-plugins/peer-forwarder/src/main/java/com/amazon/dataprepper/plugins/prepper/peerforwarder/discovery/AwsCloudMapPeerListProvider.java [hunk 1]
```diff
@@ -106,6 +113,16 @@ private static String getRequiredSettingString(final PluginSetting pluginSetting
         return Objects.requireNonNull(propertyValue, String.format("Missing '%s' configuration value", propertyName));
     }
 
+    private static Map<String, String> getOptionalSettingMap(final PluginSetting pluginSetting, final String propertyName) {
+        final Map<String, String> propertyValue = pluginSetting.getTypedMap(propertyName, String.class, String.class);
+
+        if (propertyValue == null) {
```
Comment (cmanning09) [inline data-prepper-plugins/peer-forwarder/src/main/java/com/amazon/dataprepper/plugins/prepper/peerforwarder/discovery/AwsCloudMapPeerListProvider.java:119]: Out of scope for PR: We should eliminate null checks by enforcing a default value in the PluginSettings or allow calling classes to specify the default. Raising: #1563

**Review by cmanning09 — APPROVED** [review]

File: data-prepper-plugins/peer-forwarder/src/test/java/com/amazon/dataprepper/plugins/prepper/peerforwarder/discovery/AwsCloudMapPeerListProviderTest.java [hunk 2]
```diff
@@ -65,6 +67,7 @@ void setUp() {
         awsServiceDiscovery = mock(ServiceDiscoveryAsyncClient.class);
         namespaceName = RandomStringUtils.randomAlphabetic(10);
         serviceName = RandomStringUtils.randomAlphabetic(10);
+        queryParameters = Collections.emptyMap();
```
Comment (dlvenable) [inline data-prepper-plugins/peer-forwarder/src/test/java/com/amazon/dataprepper/plugins/prepper/peerforwarder/discovery/AwsCloudMapPeerListProviderTest.java:70]: This is using as input the default value for the query parameters. So the tests below are not verifying that that the `queryParameters` is being used. You should generate a more random map.

Relatedly, we should have some tests which explicitly test when the value is `null`.

**Review by dlvenable — CHANGES_REQUESTED** [review]

File: data-prepper-plugins/peer-forwarder/src/test/java/com/amazon/dataprepper/plugins/prepper/peerforwarder/discovery/AwsCloudMapPeerListProviderTest.java (hunk 2 above)
Comment (engechas) [reply data-prepper-plugins/peer-forwarder/src/test/java/com/amazon/dataprepper/plugins/prepper/peerforwarder/discovery/AwsCloudMapPeerListProviderTest.java:70]: Good point, I will generate a random map.

`constructor_throws_with_null_QueryParameters()` is explicitly testing for this value being null. Or are you referring to adding a test for when `pluginSetting.getTypedMap` returns null?

File: data-prepper-plugins/peer-forwarder/src/test/java/com/amazon/dataprepper/plugins/prepper/peerforwarder/discovery/AwsCloudMapPeerListProvider_CreateTest.java [hunk 3]
```diff
@@ -72,6 +75,18 @@ void createPeerListProvider_with_missing_required_property(final String property
...
     }
 
+    @ParameterizedTest
+    @ValueSource(strings = {
+            PeerForwarderConfig.AWS_CLOUD_MAP_QUERY_PARAMETERS
+    })
+    void createPeerListProvider_with_missing_optional_map_property(final String propertyToRemove) {
+        pluginSetting.getSettings().remove(propertyToRemove);
+
+        final PeerListProvider result = AwsCloudMapPeerListProvider.createPeerListProvider(pluginSetting, pluginMetrics);
```
Comment (dlvenable) [inline data-prepper-plugins/peer-forwarder/src/test/java/com/amazon/dataprepper/plugins/prepper/peerforwarder/discovery/AwsCloudMapPeerListProvider_CreateTest.java:85]: Thanks! This is the bit of code that I was hoping to test.

The only missing part is to verify that a subsequent call to the AWS SDK has an empty map.

Comment (engechas) [reply data-prepper-plugins/peer-forwarder/src/test/java/com/amazon/dataprepper/plugins/prepper/peerforwarder/discovery/AwsCloudMapPeerListProvider_CreateTest.java:85]: The `ServiceDiscoveryAsyncClient` is created as a local variable in the `createPeerListProvider` so there's not a good way to mock it in this flow without introducing PowerMock as a new dependency or a decent chunk of refactoring.

I will drop the private scope from the `getOptionalSettingMap` to test the various scenarios while keeping this PR minimal.

**Review by dlvenable — APPROVED** [review]

**Review by cmanning09 — APPROVED** [review]

**Pull request merged**